### ISSUE TITLE
prometheus operator cant monitor other namespaces

### ISSUE THREAD

**Issue by zfrhv** [prometheus-operator/prometheus-operator#4601]
**What happened?**

installed prometheus certified operator 0.47.0 on openshift 4.8.19
configured for the prometheus to monitor all namespaces:
```
apiVersion: monitoring.coreos.com/v1
kind: Prometheus
metadata:
  name: test
  namespace: monitoring
spec:
  serviceMonitoringSelector: {}
  serviceMonitoringNamespaceSelector: {}
  ruleSelector: {}
  ruleNamespaceSelector: {}
  serviceAccountName: prometheus-k8s
```

created test rule in current namespace and in different namespace, it detects the rule in current namespace but not in the other namespace.

i gave cluster-admin to both prometheus-operator and prometheus-k8s service accounts.

no any error or strange logs in the operator pod.

only 1 strange thing, when the prometheus pod starts, it fails with error:
`level=error msg="Error loading config (--config.file=/etc/prometheus/config_out/prometheus.env.yaml)" err="open /etc/prometheus/config_out/prometheus.env.yaml: no such file or directory"`
but then after first restart it runs successfully without any errors

(and i also work in airgapped network)

**Did you expect to see something different?**

to see both prometheusRules in prometheus ui

**How to reproduce it (as minimally and precisely as possible)**:

just install prometheus operator beta 0.47.0 on openshift 4.8.19, always happens

**Environment**

ocp 4.8.19

* Prometheus Operator version:

    `Insert image tag or Git SHA here`
    
    
tag: v0.47.0

* Kubernetes version information:
kubernetes version: v1.21.4+6438632
    

* Kubernetes cluster kind:

    installed openshift upi on vmware

* Manifests:

```
insert manifests relevant to the issue
```

* Prometheus Operator Logs:

```
insert Prometheus Operator logs relevant to the issue here
```
sorry i dont want to hand write all the logs, getting it out from airgapped network is pain.

**Anything else we need to know?**:
i would be happy if anyone could tell how to run the operator pod in debug mode.

the selection from other namespaces does works for prometheusRule, nor serviceMonitors

**Comment by simonpasquier** [comment]
Not related to Prometheus rules but `serviceMonitoringSelector` and `serviceMonitoringNamespaceSelector` aren't valid fields.

>  i would be happy if anyone could tell how to run the operator pod in debug mode.

add `--log-level=debug` to the container args in the operator's deployment manifest. 

> sorry i dont want to hand write all the logs, getting it out from airgapped network is pain.

well it's going to be difficult for anybody to troubleshoot.

**Comment by zfrhv** [comment]
> add --log-level=debug to the container args in the operator's deployment manifest.

thanks,
well i ran the operator pod in debug (changed the csv),
and looks like the operator sends GET requests to list resources only for the namespace which its installed on (monitoring).
could it be because i installed the operator on a spesific namespace and not "All namespaces on the cluster" ? (this option is not available when i install the operator)

---
> Not related to Prometheus rules but serviceMonitoringSelector and serviceMonitoringNamespaceSelector aren't valid fields.

can i know why they are not valid fields? looks like [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/Documentation/user-guides/getting-started.md#include-servicemonitors) they do use `serviceMonitorSelector`

---
> well it's going to be difficult for anybody to troubleshoot.

yea sorry, i can get the logs out but its difficult proccess. if we will be really stuck then i will do it

thanks again ~

**Comment by simonpasquier** [comment]
> could it be because i installed the operator on a spesific namespace and not "All namespaces on the cluster" ? (this option is not available when i install the operator)

probably yes.

**Comment by zfrhv** [comment]
i saw [here](https://github.com/prometheus-operator/prometheus-operator/blob/ed36ac53f3ee2635618459cb91291b6a739b9199/pkg/listwatch/namespace_denylist_test.go#L116) that namespace `monitoring` is some kind of special or something,
so i tried to install it in namespace `ocp-monitoring`, and still have the same issue :/

**Comment by zfrhv** [comment]
sorry my bad, its not certified operator its community operator.

anyway i tried to install [kube-prometheus-stack](https://artifacthub.io/packages/helm/prometheus-community/kube-prometheus-stack) helm chart, and it was conflicting with the openshift-monitoring application (both of them ran into crash look back off after the installation)

so what should i do?
i cant install the operator in all namespaces because this option is just not available.

i may be some other problem, but i cant find any error.

is it working for anyone else?
do other people also cant monitor outside the namespace or only i have this problem?

**Comment by zfrhv** [comment]
well well,
i installed the operator by just applying the bundle `oc apply -f bundle.yaml`.

but seems like he started to manage all of the instances in all namespaces (which is good?),
including the default namepaces as `openshift-monitoring` and `openshift-user-workload-monitoring` (which is bad).
plus the operator was in crash loop backoff without any errors. (i guess he just conflicts with the openshift-monitoring prometheus operator)

maybe thats why the option "install on all namespaces" wasnt available. because then the operator conflicts with other operator from `openshift-monitoring`.

would it be possible for the operator to be installed "on all namespaces" (i guess by doing `oc apply -f bundle.yaml`), but he would skip and not watch some namespaces that i tell him ? (for example skip the default openshift namespaces)
i think the operator maybe already skips the namespace `kube-system`, so would it be possible to make that list more generic that users could modify?

or maybe be able to install the operator on a spesific namespace, and then telling him what namespaces to monitor by modifying operator group ?

or maybe create another crd as `namespaceSelector` or `prometheusNamespaceSelector`, that will tell the operator in which namespaces to manage prometheus instances (so that he wouldnt touch prometheus instances in openshift namespaces).

thanks

**Comment by zfrhv** [comment]
or maybe just from the beginning tell that the operator will watch instances only with label of the operator :l

for example if i want the prometheus to be created by operator from namespace `test`, and on the custom resource of the prometheus i will add label of the operator.
like:
```yaml
kind: prometheus
metadata:
  labels:
    operator-namespace: test
```

or maybe instead of namespace the operator name:
```yaml
kind: prometheus
metadata:
  labels:
    operator-name: test-operator
```

**Comment by zfrhv** [comment]
so i guess currently this is not possible to have 2 operators in the same cluster? because they both try to manage prometheus instances + thanos ruler + alermanager.
or am i missing something?

**Comment by fpetkovski** [comment]
In OCP you can use UWM for another prometheus operator instance. Installing it through the bundle is not recommended nor supported. In any case, this is more of a question for OCP

**Comment by zfrhv** [comment]
yes but in user workload its very limited, the idea is that i can manage it by my own but its not the case, i cant configure whatever i want.

for example i had to monitor applications outside the cluster, but i cant do that, not with openshift-monitoring nor user workload. (i couldnt configure additional scrape configuration)

so i opened case on redhat, and they are not interested in supporting monitoring outside the cluster, and told me to install separate Prometheus for that.

so here i am ended up installing the operator, and now if i already use the operator then why do i need user workload monitoring.

**Comment by zfrhv** [comment]
i see,
so i have no other choice but for the operator to watch only hardcoded list of namespaces, or all namespaces (not labes selection)

which means that there can be only 1 operator that will watch all namespaces. and redhat already made it and its openshift-monitoring.

so i have no other choice but to use openshift-monitoring to monitor all applications, and then use thanos ruler or whatever to use those metrics.

if the operator could select in which namespaces to manage prometheus + alertmanager + thanosruler by labels it would be perfect.
but i guess this is a lot of work for a small profit.

well then i will go and use the user workload (because its already configured with openshfit-monitoring) + single Prometheus in single namespace that will only monitor outside the cluster.
not comfortable because now i need to make sure to upgrade them equally, but seems like yet its the best solution.

thank you a lot !

**Comment by zfrhv** [comment]
wait, sorry for being annoying, but the original problem is still not solved.
operator doesnt detects instances outside of his namespace.

i checked my operator deployment:
```yaml
...
      spec:
        containers:
        - args:
          - -namespaces=$(NAMESPACES)
          - --prometheus-config-reloader=quay.io/prometheus-operator/prometheus-config-reloader:0.47.0
...
```
i think the problem here is that when you spepcify `--namespace=` then the operator will read instances only from those namespaces. which makes the `serviceMonitorNamespaceSelector`, `ruleNamespaceSelector`.. useless because the operator doesnt even watches other namespaces.

i checked the code a little and seems like when you use `--namespaces` it doesnt even creates `NamespaceInformer`,  i guess it means that operator wont read those namespaces.

so i think the fix would be for the operator to read all namespaces, and then check if the instnace is in the list of `--namesapces`

looks confusing but,
for the operator to read rules from other namespaces it needs `NamespaceInformer` for all namespaces (means operator will monitor all namespaces).
- then if we will use: `ThanosRuler`, `AlertManager`, `Prometheus`, `ServiceMonitorNamespaceSelector`, `RuleNamespaceSelector`... we will need to filter out with the `Namespaces.AllowList`
- and the `serviceMonitorSelector`, `ruleNamespaceSelector`... will use his own filter specified in `ServiceMonitorNamespaceSelector`, `RuleNamespaceSelector`...

i would happily create a pull request but its my first time seeing go language. plus not knowing the structure of the operator makes it a lot more difficult :(

**Comment by zfrhv** [comment]
> Not really. You can start the operator with -namespaces=foo,bar,... and still want to apply namespace selectors for monitoring resources. The reason for -namespaces is when the operator's service account doesn't have list/watch permissions on all namespaces.

As i understand the namespace selector will work only on the list that is in the `-namespaces` flag, is that correct?
for example if the list is `-namespaces=ns1,ns2,ns3`, then the namespace selector will work only on ns1, ns2 and ns3.
the operator cant monitor anything in ns4, ns5... is that right?

when installing the community operator on a spesific namespace (the only option available), the its `-namespaces` flas is configured only on the current namespace.
i cant check it right now but if i remember right, the `$(NAMESPACES)` in env is only the namespace where the operator is installed.
so the operator cant monitor outside of its namespace?

---
> I'm not sure where you found this.

sorry maybe im wrong, im not that good in go.
but [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/cmd/operator/main.go#L163) namespaces saved into `ns`, then [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/cmd/operator/main.go#L228) its saved as `Namespaces.AllowList` in the configuration, and [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/pkg/prometheus/operator.go#L307) the `NamespaceInformers` are created from the original namespaces list.
its the only area where the function `newNamespaceInformer` was used, so i assumed that the operator doesnt even creates informer/listener for other namespaces, thus cant even monitor them.

**Comment by zfrhv** [comment]
I tried to replace this:
```yaml
        - -namespaces=$(NAMESPACES)
```
with this:
```yaml
        - --prometheus-instance-namespaces=$(NAMESPACES)
        - --alertmanager-instance-namespaces=$(NAMESPACES)
        - --thanos-ruler-instance-namespaces=$(NAMESPACES)
```
as mentioned [here](https://github.com/prometheus-operator/prometheus-operator/issues/4601#issuecomment-1058979958) (in the operator deployment by changing the operator csv), and i get next errors:
```log
level=error msg="get namespace to enqueue Prometheus instances failed: namespace \"openshift-monitoring\" doesnt not exist"
level=error msg="get namespace to enqueue Prometheus instances failed: namespace \"openshift-user-workload-monitoring\" doesnt not exist"
```
the operator pod is in CrashLoopBackOff, buttttt the prometheus pod finally detects rules from other namespaces.

the `$(NAMESPACES)` is pointing to the `metadata.annotations['olm.targetNamespaces']`, and the annotation `olm.targetNamespaces` is `"monitoring"` (the current namespace where i installed the operator).

so basically its:
```yaml
        - -namespaces=monitoring
```
which i guess thats why the operator detects all instances only in his own namespace (when using `-namespaces`).

i just dont understand few things:
- why its `-namespaces=$(NAMESPACES)` and not `-namespaces=${NAMESPACES}` ?
- why the `--prometheus-instance-namespaces=$(NAMESPACES)...` makes problems with finding `openshift-monitoring` and `openshift-user-workload-monitoring` namespaces? is it because there are prometheus instances that other operator already manages ?

**Comment by zfrhv** [comment]
oh and the CrashLoopBackOff seems to be caused by `OOMKilled`.
even after increasing the limit i was still getting OOMKilled and CrashLoopBackOff.
default resources:
```yaml
resources:
  limits:
    cpu: 200m
    memory: 200Mi
  requests:
    cpu: 100m
    memory: 100Mi
```
tried to increase to:
```yaml
resources:
  limits:
    cpu: 400m
    memory: 1024Mi
  requests:
    cpu: 100m
    memory: 100Mi
```

**Comment by zfrhv** [comment]
finally !
I just increased the limit even more and everything started to work.

I increased it to 8 cpu and 8Gb ram, then check the actual usage (with grafana),
and the operator used 0.1 cpu and 2.1 Gb memory.

so i change the limit:
```yaml
resources:
  limits:
    cpu: 300m
    memory: 3Gi
  requests:
    cpu: 100m
    memory: 100Mi
```


I want to create PR but I dont see the same configuration as my operator was installed with.
- the resources limits I found [here](https://github.com/prometheus-operator/prometheus-operator/blob/main/jsonnet/prometheus-operator/prometheus-operator.libsonnet#L10)
- but the args are different, its not `-namespace=` its `--kubelet-service=`.
I want to apply as i did [here](https://github.com/prometheus-operator/prometheus-operator/issues/4601#issuecomment-1062755061), but in todays sources its [with the `--kubelet-service`](https://github.com/prometheus-operator/prometheus-operator/blob/main/jsonnet/prometheus-operator/prometheus-operator.libsonnet#L129-L132)

so im not sure how to modify the args
